File: Civitas ruby/lib/casilla.rb
require_relative "tipo_casilla"

module Civitas
class Casilla
  @@carcel = 5
  
  def initialize(nombre, titulo, importe, numCasilla, mazo, tipo)
    @nombre = nombre
    @mazo = mazo
    @importe = importe
    @tipo = tipo
    @sorpresa = nil
    
    if(tipo == Tipo_casilla::CALLE)
      @tituloPropiedad = titulo

    end
  end
  
  def self.nombre(nombre)
    self.new(nombre, nil, nil, @@carcel, nil, Tipo_casilla::DESCANSO)
  end
  
  def self.titulo(titulo)
    self.new(titulo.getNombre, titulo, nil, @@carcel, nil, Tipo_casilla::CALLE)
  end
  
  def self.cantidad(cantidad, nombre)
    self.new(nombre, nil, cantidad, @@carcel, nil, Tipo_casilla::IMPUESTO)
  end
  
  def self.casillaCarcel(numCasillaCarcel, nombre)
    self.new(nombre, nil, nil, numCasillaCarcel, nil, Tipo_casilla::JUEZ)
  end
  
  def self.mazo(mazo, nombre)
    self.new(nombre, nil, nil, @@carcel, mazo, Tipo_casilla::SORPRESA)
  end
  
  
  
  
  
  private
  
  def recibeJugador_calle(iactual, todos)
    if(jugadorCorrecto(iactual, todos))
      informe(iactual, todos)
      jugador = todos[iactual]
      
      if(!@tituloPropiedad.tienePropietario)
        jugador.puedeComprarCasilla
      else
        @tituloPropiedad.tramitarAlquiler(jugador)
        
        if(@tituloPropiedad.tienePropietario)
          if(!@tituloPropiedad.esEsteElPropietario(jugador))
            precio = @tituloPropiedad.getPrecioAlquiler
            jugador.pagaAlquiler(precio)
            @tituloPropiedad.getPropietario.recibe(precio)
          end
        end
      end
    end
  end
  
  def recibeJugador_impuesto(iactual, todos)
    if (jugadorCorrecto(iactual, todos))
      informe()
      todos[iactual].pagaImpuesto(@impuesto)
    end
  end
  
  def recibeJugador_juez(iactual, todos)
    if (jugadorCorrecto(iactual, todos))
      informe(iactual, todos)
      todos[iactual].encarcelar(@@carcel)
    end
  end
  
  def recibeJugador_sorpresa(iactual, todos)
    if(jugadorCorrecto(iactual, todos))
      sorpresa = @mazo.siguiente()
      informe(iactual, todos)
      sorpresa.aplicarAJugador(iactual, todos)
    end
  end
  
    
  def informe(iactual, todos)
    Diario.instance.ocurre_evento("\nEl jugador " + todos[iactual].getNombre + " cae en" + toString())
  end
  
  
  
  
  public
  
  def getNombre()
    return @nombre
  end 
  
    
  def jugadorCorrecto(iactual, todos)
    if(iactual < todos.length)
      return true
    else
      return false
    end
  end
  
    
  def toString()
    cadena = "\nNombre: " + @nombre +
      "\nTipo: " + @tipo.to_s
    
    return cadena
  end
  
  def getTituloPropiedad
    return @tituloPropiedad
  end
  
  def recibeJugador(iactual, todos)
    if(@tipo == Tipo_casilla::CALLE)
      recibeJugador_calle(iactual, todos)
    elsif (@tipo == Tipo_casilla::IMPUESTO)
      recibeJugador_impuesto(iactual, todos)
    elsif (@tipo == Tipo_casilla::JUEZ)
      recibeJugador_juez(iactual, todos)
    elsif (@tipo == Tipo_casilla::SORPRESA)
      recibeJugador_sorpresa(iactual, todos)
    else
      informe(iactual, todos)
    end
  end
end
end

File: Civitas java/src/civitas/Casilla.java
package civitas;

import java.util.ArrayList;

public class Casilla {
    private static int carcel;
    private float importe;
    private String nombre;
    private TipoCasilla tipo;
    private TituloPropiedad tituloPropiedad;
    private MazoSorpresas mazo;
    private Sorpresa sorpresa;
    
    
    Casilla(String nombre1){
        init();
        nombre = nombre1;
        tipo = TipoCasilla.DESCANSO;
    }
    
    Casilla(TituloPropiedad titulo){
        init();
        nombre = titulo.getNombre();
        tipo = TipoCasilla.CALLE;
        tituloPropiedad = titulo;
    }
    
    Casilla(float cantidad, String nombre1){
        init();
        nombre = nombre1;
        importe = cantidad;
        tipo = TipoCasilla.IMPUESTO;
    }
    
    Casilla(int numCasillaCarcel, String nombre1){
        init();
        nombre = nombre1;
        tipo = TipoCasilla.JUEZ;
        carcel = numCasillaCarcel;
    }
    
    Casilla(MazoSorpresas mazo1, String nombre1){
        init();
        nombre = nombre1;
        mazo = mazo1;
        tipo = TipoCasilla.SORPRESA;
    }
    
    public String getNombre(){
        return nombre;
    }
    
    TituloPropiedad getTituloPropiedad(){
        return tituloPropiedad;
    }
    
    private void informe(int iactual, ArrayList<Jugador> todos){
        String nombre = todos.get(iactual).getNombre();
        String cadena = toString();
        
        Diario.getInstance().ocurreEvento("\nEl jugador " + nombre + " ha caído en: \n" + cadena);
    }
    
    private void init(){
        importe = 0;
        mazo = null;
        sorpresa = null;
        tituloPropiedad = null;
        tipo = null;
        nombre = "";
    }
    
    public boolean jugadorCorrecto(int iactual, ArrayList<Jugador> todos){
        if(iactual < todos.size())
            return true;
        else
            return false;
    }
    
    void recibeJugador(int iactual, ArrayList<Jugador> todos){
        if(tipo == TipoCasilla.CALLE)
            recibeJugador_calle(iactual, todos);
        else if(tipo == TipoCasilla.IMPUESTO)
            recibeJugador_impuesto(iactual, todos);
        else if(tipo == TipoCasilla.JUEZ)
            recibeJugador_juez(iactual, todos);
        else if(tipo == TipoCasilla.SORPRESA)
            recibeJugador_sorpresa(iactual, todos);
        else
            informe(iactual, todos);
    }
    
    private void recibeJugador_calle(int iactual, ArrayList<Jugador> todos){
        if(jugadorCorrecto(iactual, todos)){
            informe(iactual, todos);
            Jugador jugador = todos.get(iactual);
            
            if(!tituloPropiedad.tienePropietario()){
                jugador.puedeComprarCasilla();
            } else {
                tituloPropiedad.tramitarAlquiler(jugador);
                
                if(tituloPropiedad.tienePropietario()){
                    if(tituloPropiedad.esEsteElPropietario(jugador)){
                        jugador.pagaAlquiler(tituloPropiedad.getPrecioAlquiler());
                        tituloPropiedad.getPropietario().recibe(tituloPropiedad.getPrecioAlquiler());
                    }
                }
            }
        }
    }
    
    private void recibeJugador_impuesto(int iactual, ArrayList<Jugador> todos){
        if(jugadorCorrecto(iactual, todos)){
            informe(iactual, todos);
            todos.get(iactual).pagaImpuesto(importe);
        }
    }
    
    private void recibeJugador_juez(int iactual, ArrayList<Jugador> todos){
        if(jugadorCorrecto(iactual, todos)){
            informe(iactual, todos);
            todos.get(iactual).encarcelar(carcel);
        }
    }
    
    private void recibeJugador_sorpresa(int iactual, ArrayList<Jugador> todos){
        if(jugadorCorrecto(iactual, todos)){
            Sorpresa sorpresa = mazo.siguiente();
            informe(iactual, todos);
            sorpresa.aplicarAJugador(iactual, todos);
        }
    }
    
    public String toString(){
        String cadena = "\nNombre: " + nombre +
            "\nTipo: " + tipo + "\n\n";
        
        return cadena;
    }
}
